[New PowerToy] File Locksmith

ivanstosic-janea opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->

The old PR where discussion took place: #20863 

## Summary of the Pull Request

This PR adds a new PowerToy that can be used to check which processes are holding files open, thus preventing them from being deleted, for example. It does this by adding a context menu entry in the Windows Explorer, activated for all files and directories.

This PR is a draft, and most likely the UI needs some improvement.

![file-locksmith](https://user-images.githubusercontent.com/82372754/193033091-2f1f9b83-eef8-44bb-8e9a-a825fe5c2282.gif)

## Things that need to be done:

* A few strings in the GUI should be localized. I think I did localize the strings in the context menu entry, but not sure if I did it correctly.
* Consider adding some settings. I couldn't come up with anything.
* Support Windows themes (Light/Dark)
* Support GUI restarting itself as elevated. It's not possible to see handles held by elevated processes, so some results will be missing in that case, unless the GUI itself is elevated.

<!-- Please review the items on the PR checklist before submitting-->

## Development and testing

I modified the installer script to add registry entries, so this utility should work out of the box. If you plan to do development on it, there is a more convenient way to test it. Build the solution, run `cmd` and go to the directory with the output binaries, then run `regsvr32 PowerToys.FileLocksmithExt.dll`. This will call `DllRegisterServer` and that will add the necessary registry keys. If you want to remove these registry keys, run `regsvr32 /u PowerToys.FileLocksmithExt.dll`.

After opening the context menu in Windows Explorer, the dll will be loaded and Explorer will prevent you from overwriting it. In that case you can just kill `explorer.exe` and restart it. The dll will not be held until you open the context menu again, allowing you to recompile the dll.

## PR Checklist

- [x] **Closes:** #114
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

ivanstosic-janea committed:
Imported offline solution

ivanstosic-janea committed:
Make solution compile

ivanstosic-janea committed:
Add Windows sample, doesn't work?

ivanstosic-janea committed:
Added new project to implement the dll

ivanstosic-janea committed:
Remove unneeded header

ivanstosic-janea committed:
Implemented IUnknown part of ExplorerCommand

ivanstosic-janea committed:
Implemented IExplorerCommand methods

ivanstosic-janea committed:
Implemented ClassFactory

ivanstosic-janea committed:
Implemented DLL register/unregister

ivanstosic-janea committed:
Implemented other DLL exports, not working?

ivanstosic-janea committed:
Implemented IShellExtInit inferface

ivanstosic-janea committed:
Implemented IContextMenu, it works!

ivanstosic-janea committed:
Implement command data fetching

ivanstosic-janea committed:
Make sample project compile on VS 2022

ivanstosic-janea committed:
Add plan

ivanstosic-janea committed:
Implement Lib as separate project

ivanstosic-janea committed:
Implemented IPC, not tested

ivanstosic-janea committed:
Console UI project skeleton

ivanstosic-janea committed:
Implemented basic console UI

ivanstosic-janea committed:
Implemented piping, there are bugs

ivanstosic-janea committed:
Prototype works

ivanstosic-janea committed:
Remove old project

ivanstosic-janea committed:
Added GUI project skeleton

ivanstosic-janea committed:
Mitigate issue with WinUI3

ivanstosic-janea committed:
Added a control for displaying results

ivanstosic-janea committed:
Add button

ivanstosic-janea committed:
Implement core functions in lib project

ivanstosic-janea committed:
Call new library function from console main

ivanstosic-janea committed:
Implement showing results

ivanstosic-janea committed:
Improve UI

ivanstosic-janea committed:
Implemented subdirectory search

ivanstosic-janea committed:
Remove useless code

ivanstosic-janea committed:
Set window size

ivanstosic-janea committed:
UI adjustments

ivanstosic-janea committed:
Implement killing process

ivanstosic-janea committed:
Rename variable

ivanstosic-janea committed:
Add lib project to main solution

ivanstosic-janea committed:
Add Ext and GUI projects to solution

ivanstosic-janea committed:
Tweak packages for GUI project

ivanstosic-janea committed:
Add a settings page

ivanstosic-janea committed:
Add a few resource strings

ivanstosic-janea committed:
Add one more resources string

ivanstosic-janea committed:
VS keeps trying to correct this

ivanstosic-janea committed:
Add references to File Locksmith in /,github

ivanstosic-janea committed:
Implement some parts of FileLocksmithModule

ivanstosic-janea committed:
Change output directory

ivanstosic-janea committed:
Change target name and add to runner

ivanstosic-janea committed:
Add logger

ivanstosic-janea committed:
Started implementing settings backend

ivanstosic-janea committed:
Fix log folder

ivanstosic-janea committed:
Settings work

ivanstosic-janea committed:
Add some basic tracing

ivanstosic-janea committed:
Attempt at adding resources

ivanstosic-janea committed:
Remove junk files

ivanstosic-janea committed:
Added missing defines

ivanstosic-janea committed:
Replaced some constants with resources

Something's not working

ivanstosic-janea committed:
Move resources to the Ext project

ivanstosic-janea committed:
Remove experiment

ivanstosic-janea committed:
Add binaries for signing

ivanstosic-janea committed:
Improve tracing

ivanstosic-janea committed:
Remove old Settings calls

ivanstosic-janea committed:
Show something when there are no results

ivanstosic-janea committed:
Change window title

ivanstosic-janea committed:
Move computation to another thread, improve UX

ivanstosic-janea committed:
Increase font size for default text

ivanstosic-janea committed:
Remove entries for killed processes

ivanstosic-janea committed:
Show user name

ivanstosic-janea committed:
Remove nonrecursive implementation

ivanstosic-janea committed:
Implement back end for getting file names

ivanstosic-janea committed:
Show list of files, UI tweaks

ivanstosic-janea committed:
Remove useless includes

ivanstosic-janea committed:
Implement back end for getting full process path

ivanstosic-janea committed:
Dark title bar on dark themes

ivanstosic-janea committed:
Using Expander, other UI adjustments

ivanstosic-janea committed:
Show "No results" after killing all processes

ivanstosic-janea committed:
Show progress ring

ivanstosic-janea committed:
Update configuration mapping

htcfreek commented:
I personally prefer the ui from @niels9001's mock-up. Do we switch to this ui?

ivanstosic-janea committed:
Revert "Update configuration mapping"

This reverts commit d8e13206f3c7de3c6dbf880299bfff3bf9f27a37.

ivanstosic-janea commented:
> I personally prefer the ui from @niels9001's mock-up. Do we switch to this ui?

No, that is still the old UI. If I understood it correctly, Neils will push changes to this branch.

ivanstosic-janea committed:
Fixed solution configuration, ARM64 should build

ivanstosic-janea committed:
Backend for refreshing

ivanstosic-janea committed:
Variable window size

ivanstosic-janea committed:
Add refresh button

ivanstosic-janea committed:
New WinUI3 C# project for FL

ivanstosic-janea committed:
Started porting functionality

ivanstosic-janea committed:
Add Interop project

ivanstosic-janea committed:
Move IPC to Ext project

ivanstosic-janea committed:
Ported native functions to Interop

ivanstosic-janea committed:
Ported finding processes

ivanstosic-janea committed:
Ported most of Main Window functionality

ivanstosic-janea committed:
Display paths of files

ivanstosic-janea committed:
Implement killing processes

ivanstosic-janea committed:
Use resource string for "End Task"

ivanstosic-janea committed:
Remove entries for terminated processes

ivanstosic-janea committed:
Show User name

ivanstosic-janea committed:
Set default window size

ivanstosic-janea committed:
Make the new UI the default

ivanstosic-janea committed:
Reading paths from stdin, completed port to C#

ivanstosic-janea committed:
Fix small bug

Wolverine1977 commented:
What version is this planned to be included in?

niels9001 committed:
Moving to MVVM

htcfreek commented:
@ivanstosic-janea 
How does the ui looks at the moment? Is the screenshot still up to date?

niels9001 commented:
> @ivanstosic-janea 
> How does the ui looks at the moment? Is the screenshot still up to date?

@htcfreek @ivanstosic-janea FYI, I'll start pushing my UI changes later today and update the screenshot👍

niels9001 committed:
Adding Labs

niels9001 committed:
Reintroducing features

niels9001 committed:
Moving UI strings to resources file

crutkas commented:
What is needed to merge this in (and it has conflicts)

ivanstosic-janea commented:
I can't build it now:

`error NU1101: Unable to find package CommunityToolkit.Labs.WinUI.SettingsControls. No packages exist with this id in source(s): Microsoft Visual Studio Offline Packages, nuget.org`

htcfreek commented:
> I can't build it now:
> 
> `error NU1101: Unable to find package CommunityToolkit.Labs.WinUI.SettingsControls. No packages exist with this id in source(s): Microsoft Visual Studio Offline Packages, nuget.org`

@ivanstosic-janea 
Maybe you have to update your nuget source: https://devblogs.microsoft.com/ifdef-windows/announcing-the-windows-community-toolkit-labs/#using-labs-experiments-in-your-app

ivanstosic-janea commented:
Can this somehow be automated, or does everyone working on PowerToys have to do this manually? In that case this must be documented somewhere.

niels9001 commented:
> Can this somehow be automated, or does everyone working on PowerToys have to do this manually? In that case this must be documented somewhere.

Good and fair question, will check this with the team!

htcfreek commented:
> Can this somehow be automated, or does everyone working on PowerToys have to do this manually? In that case this must be documented somewhere.

And does this have impacts on our build farm? Do we have to adjust the configuration before merging this PR?

ivanstosic-janea commented:
I just checked the UI now, it looks awesome, thanks for the great work!

Tomorrow I will add back one feature - when a listed process exits, it should be automatically removed from the results list - whether or not it was killed from FL or externally.

htcfreek commented:
> I just checked the UI now, it looks awesome, thanks for the great work!

Can you update the gif in the PR description. I am interested in how it looks now.

ivanstosic-janea commented:
I will not yet replace the GIF in the PR description because some old features are now missing. But this is what it looks like:
 
![file-locksmith-niels](https://user-images.githubusercontent.com/82372754/196059202-8c636002-4d76-402f-9217-db768fa9824d.gif)

ivanstosic-janea committed:
Restored functionality

htcfreek commented:
@ivanstosic-janea 
I think we should have a small header at the top of the window. I think something like "Results for [folder/file name]:" for single explorer item selection and something like "Results for [folder/file name] and [X] more items:" for multiple explorer items. On multiple explorer items we can show an information symbol with tooltip to list all selections. 

I think this is helpful when working with many windows.

ivanstosic-janea commented:
@niels9001 I can't get the installer to work, are there any files that need to be added other than `CommunityToolkit.Labs.WinUI.SettingsControls.dll`?

niels9001 commented:
> @niels9001 I can't get the installer to work, are there any files that need to be added other than `CommunityToolkit.Labs.WinUI.SettingsControls.dll`?

Specific errors that you're getting?

ivanstosic-janea commented:
When you build the PT installer, install PT that way and run FileLocksmithUI, it crashes on launch. Probably some file is missing from the installer script.

niels9001 commented:
> When you build the PT installer, install PT that way and run FileLocksmithUI, it crashes on launch. Probably some file is missing from the installer script.

installer/PowerToysSetup/Product.wxs should all nuget packages/dlls be added here as well? I'm not familiar with how the installer works exactly, maybe @stefansjfw or @yuyoyuppe can chime in on how to get a proper error?

yuyoyuppe commented:
you could run the `.dll` version of .exe with `dotnet` cmd and it'll show you which dll is missing.

ivanstosic-janea committed:
Add missing dlls

ivanstosic-janea commented:
Looks like we're also missing one DLL in the build output `modules/FileLocksmith/runtimes/` folder. Oh, and I found it by running File Locksmith :)

ivanstosic-janea commented:
/azp run

ivanstosic-janea committed:
Add FIle Locksmith to publish.cmd

ivanstosic-janea committed:
Merge branch 'ivan/file-locksmith' of https://github.com/microsoft/PowerToys into ivan/file-locksmith

Removing one parent commit for cleaner history

Co-Authored-By: Niels Laute <niels.laute@live.nl>

ivanstosic-janea committed:
Rebase fixes

ivanstosic-janea committed:
Try updating nuget.config

ivanstosic-janea committed:
Fix copy-paste blunder

ivanstosic-janea committed:
Add File Locksmith UI for publishing

ivanstosic-janea committed:
Add .pubxml file in FileLocksmith

ivanstosic-janea committed:
Change build output folder

jaimecbernardo committed:
Fix installer build issues

Remove old projects from solution so MSBuild doesn't build them.
Downgrade target framework to what most other projects are using.
Fix publishing profile and project runtimes.
Remove unused CsWinRT references.

jaimecbernardo committed:
[CI] Add clear to nuget packages

ivanstosic-janea committed:
Fix module reference counting

jaimecbernardo committed:
Fix nuget for release CI

jaimecbernardo committed:
Fix version and signing

jaimecbernardo committed:
Fix path for resources

ivanstosic-janea committed:
Fix incorrect results when running 2 instances

jaimecbernardo commented:
Looking at the way it looks on Windows 10, I'm assuming it's using some features that are only available on Windows 11.
@niels9001 , maybe I add the fallback I've added for Hosts File Editor as well? Use the default PowerToys icon? WDYT?

![image](https://user-images.githubusercontent.com/26118718/197760941-33810683-6e9e-4b28-822f-e4ad22033448.png)

niels9001 commented:
> Looking at the way it looks on Windows 10, I'm assuming it's using some features that are only available on Windows 11. @niels9001 , maybe I add the fallback I've added for Hosts File Editor as well? Use the default PowerToys icon? WDYT?
> 

Yes please! If you can add a placeholder .ico I'll update the assets soon with a new icon.

With WASDK 1.2 the custom titlebar should work for W10 as well, all the way to 1809. According to the latest Community Call, release should be next week-ish _~ 1 Nov)!

jaimecbernardo commented:
Sounds good. Thank you!
Perhaps WASDK 1.2 release for 0.65? I think that's a bit too late for 0.64.

jaimecbernardo committed:
Fix default nuget source

jaimecbernardo committed:
Windows 10 icon and fallback for UI

jaimecbernardo commented:
Looking better.
@niels9001 , it's that `icon.ico` file in "scr/modules/FileLocksmith/FileLocksmithUI".

![image](https://user-images.githubusercontent.com/26118718/197781123-96ab1705-1f22-4431-9dec-b8f29ea2b7b8.png)

jaimecbernardo committed:
Code clean-up and spaces instead of tabs

htcfreek commented:
> @ivanstosic-janea 
> I think we should have a small header at the top of the window. I think something like "Results for [folder/file name]:" for single explorer item selection and something like "Results for [folder/file name] and [X] more items:" for multiple explorer items. On multiple explorer items we can show an information symbol with tooltip to list all selections. 
> 
> I think this is helpful when working with many windows.

@ivanstosic-janea , @niels9001 
What about this. Do you missed this?

ivanstosic-janea committed:
Showing selected paths in the header

ivanstosic-janea committed:
Add gif showcasing FL

ivanstosic-janea committed:
Add screenshot of File Locksmith for Settings

ivanstosic-janea committed:
Add new files to the installer

ivanstosic-janea commented:
@htcfreek

I have a branch with a prototype, might need some tweaking before I merge it into this one:
![image](https://user-images.githubusercontent.com/82372754/198035873-f36423a9-658e-4216-93a7-2642d3c3eda2.png)

https://github.com/ivanstosic-janea/PowerToys/tree/file-locksmith-paths-in-header

jaimecbernardo committed:
Add OOBE page

jaimecbernardo commented:
Hi @niels9001 , We were missing a fluent icon for File Locksmith, so I've just used the PowerToys default one:
![image](https://user-images.githubusercontent.com/26118718/198052036-64bbce18-d68a-4f71-848b-e71f6f463b79.png)

If you do the icon, can you please replace "src\settings-ui\Settings.UI\Assets\FluentIcons\FluentIconsFileLocksmith.png" as well? Thank you!

jaimecbernardo committed:
Merge branch 'main' into ivan/file-locksmith

ivanstosic-janea committed:
Tweak path list

ivanstosic-janea committed:
Added new, wider gif

jaimecbernardo committed:
Add GPO

jaimecbernardo committed:
Merge remote-tracking branch 'origin/ivan/file-locksmith' into ivan/file-locksmith

htcfreek commented:
@ivanstosic-janea 
Should we add an tooltip for the +X paths?

jaimecbernardo committed:
Add some logs

jaimecbernardo committed:
[CI]Get CommunityToolkit.Labs from BigPark feed

jaimecbernardo committed:
[CI]Use azure package feed for Nuget in release

jaimecbernardo committed:
[CI]Another try for the labs source

jaimecbernardo committed:
Revert changes to feed

jaimecbernardo committed:
Use RestoreAdditionalProjectSources

ivanstosic-janea committed:
Add tooltip to file list

jaimecbernardo committed:
Change tooltip to not trim the lines

htcfreek commented on src/modules/FileLocksmith/FileLocksmithUI/Views/MainPage.xaml:
```diff
@@ -0,0 +1,185 @@
+                Style="{ThemeResource BodyStrongTextBlockStyle}"
+                Text="{x:Bind ViewModel.Paths, Converter={StaticResource fileListToDescriptionConverter}}">
+                <ToolTipService.ToolTip>
+                    <TextBlock TextWrapping="WrapWholeWords" Text="{x:Bind ViewModel.PathsToString, Mode=OneWay}"/>
```
Is this tested with long paths containing spaces?
We should bind visibility on the number of paths and show it only with more than 1 path(s).

jaimecbernardo replied:
Even with 1 path, it's good to have the tooltip because the path might be too long to show in the UI.

The tooltips seem to not expand, so the previous List would not work well and would trim paths horizontally, making the information kind of useless. I've changed it to a TextBlock with text wrapping to avoid that. Here's what it looks like:
![image](https://user-images.githubusercontent.com/26118718/198262704-e3001264-0ec0-41d6-8086-412f161dc8c8.png)

It still won't show every path if there's lots of files selected, but at least it'll show useful information for long paths.

htcfreek replied:
Hmmm. But did you selected all this paths in explorer? What I had in mind was showing the items that the user has selected in explorer window (input parameters). What we have yet is a bit to much.

jaimecbernardo replied:
Yes, these were the items I selected. I opened the folder and selected all the files/folders within with select all and only then activated File Locksmith for that selection.

jaimecbernardo committed:
Add Tips and tricks section mentioning elevated

jaimecbernardo committed:
Add some more logs messages.

jaimecbernardo reviewed (approved):
There are still some todos in code, but looks good as a minimum likeable product.
When merging let's also open an issue about getting localization in for the next versions.

ivanstosic-janea committed:
Grammar fix

htcfreek commented:
I will give it a try later this day.

jaimecbernardo committed:
Add to bug report tool

ivanstosic-janea committed:
Fix UI virtualization not working

carperei commented:
can this be used for drive letters too? that would be great specially for ejecting usb drives problems.

jaimecbernardo committed:
Disable virtualization to avoid crashes

jaimecbernardo committed:
Get better virtualization

jaimecbernardo committed:
Merge branch 'main' into ivan/file-locksmith

htcfreek commented:
@jaimecbernardo 
Can you help me with these errors? Did we update winappsdk?
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	MIDL5157	[msg]An invalid winmd file was specified for import [context]: D:\AppDev\Github\PT_FileLock\packages\Microsoft.WindowsAppSDK.1.1.5\lib\uap10.0\Microsoft.Windows.ApplicationModel.WindowsAppRuntime.winmd	PowerToys.MeasureToolCore	D:\AppDev\Github\PT_FileLS\src\modules\MeasureTool\MeasureToolCore\PowerToys.MeasureToolCore.idl	1	

Severity	Code	Description	Project	File	Line	Suppression State
Error	MIDL1001	[msg]cannot open input file [context]Error opening or processing WinMD d:\appdev\github\pt_filelock\packages\microsoft.windowsappsdk.1.1.5\lib\uap10.0\microsoft.windows.system.winmd. HRESULT 0x80070003.	PowerRenameUI	D:\AppDev\Github\PT_FileLS\src\modules\powerrename\PowerRenameUILib\midlrt	1	
```

htcfreek commented on src/modules/FileLocksmith/FileLocksmithUI/ViewModels/MainViewModel.cs:
```diff
@@ -0,0 +1,162 @@
+        {
+            get
+            {
+                return string.Join("; ", paths);
```
```suggestion
                return string.Join(";\n", paths);
```

Using a line break instead of a space makes reading the paths easier.
![image](https://user-images.githubusercontent.com/61519853/198382815-815f5f10-8d9a-43ba-832f-b81894679ffc.png)

jaimecbernardo replied:
Changed to new line. THank you! ;)

htcfreek commented:
@jaimecbernardo, @ivanstosic-janea 
I tried the tool and found some things we should improve:
1. Is the reload icon in the middle of the window a button? Then it doesn't work when you open the tool from context menu with a path where no files are locked.
![image](https://user-images.githubusercontent.com/61519853/198380963-386c72c7-cafa-4b92-85ef-e2641e8f18e7.png)

2. On long paths the "+X" info isn't visible.
![image](https://user-images.githubusercontent.com/61519853/198381112-27d4c272-1604-4b26-b73c-f6698a0c23ad.png)

3. On long path the tooltip isn't wrapping correctly:
![image](https://user-images.githubusercontent.com/61519853/198381473-b27c5757-6b7c-420f-ba47-968afad3208c.png)

4. On many selected items (in my case 104) the tooltip is too small:
![image](https://user-images.githubusercontent.com/61519853/198381560-03092712-785a-4671-b2e5-f16706265037.png)

5. The context menu has no icon. We should add one. (Or is this happening because I created the entry manually.)

For the tooltip and header, I suggest showing only the name of the folder/file and not the full path. Then we might be able to use to list view with a scroll bar. Or we create our own tool tip control.

jaimecbernardo committed:
Add dialog instead of tooltip to show list of items

jaimecbernardo commented:
Hi @htcfreek, with the latest change:
- For 2: Now only the last part of the path is shown and the textbox is really a button:
![image](https://user-images.githubusercontent.com/26118718/198538386-00e575f1-b1df-42b8-a604-ae680a17f264.png)

- For 3 and 4, instead of a tooltip, clicking the header in top will now open a dialog that lists the full paths, with some scroll bars if needed:
![image](https://user-images.githubusercontent.com/26118718/198538066-e4c49b19-94af-407d-b8b9-8059d9758cae.png)

- For 1, it's not supposed to be a button. (Wouldn't be a bad idea though.)

- For 5, there's no icon yet still. I'm guessing this will be an item to follow up on later.

jaimecbernardo committed:
No results refresh icon is now a button too

jaimecbernardo commented:
And poof! Now for 1, that's a button :)

htcfreek commented:
@jaimecbernardo 
Why not making the selected folder name info directly an button or add an information button after the name in the same line. That would look cleaner.

ivanstosic-janea committed:
Use managed methods for handling processes

jaimecbernardo commented:
> @jaimecbernardo Why not making the selected folder name info directly an button or add an information button after the name in the same line. That would look cleaner.

What do you mean? That's a button.

jaimecbernardo committed:
Remove registry code from Ext.

jaimecbernardo committed:
Support drives too

jaimecbernardo commented:
> can this be used for drive letters too? that would be great specially for ejecting usb drives problems.

Drive support has been added.

htcfreek commented:
> > @jaimecbernardo Why not making the selected folder name info directly an button or add an information button after the name in the same line. That would look cleaner.
> 
> What do you mean? That's a button.

In your screenshot you have a button and the path text. We could combine it.

jaimecbernardo commented:
> > > @jaimecbernardo Why not making the selected folder name info directly an button or add an information button after the name in the same line. That would look cleaner.
> > 
> > 
> > What do you mean? That's a button.
> 
> In your screenshot you have a button and the path text. We could combine it.

That's not a button, that's the tooltip on mouse over :)

SeraphimaZykova reviewed (approved):
Tested on both platforms, works well. Code also looks good.
Great work!

joedf commented:
I'm really looking forward to this. I am wondering if there should be a "Unlock file" option (likely using [`UnlockFileEx()`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-unlockfileex)). And, perhaps even a "Kill All" option?

htcfreek commented:
> > > > @jaimecbernardo Why not making the selected folder name info directly an button or add an information button after the name in the same line. That would look cleaner.
> > > 
> > > 
> > > What do you mean? That's a button.
> > 
> > In your screenshot you have a button and the path text. We could combine it.
> 
> That's not a button, that's the tooltip on mouse over :) 

now I see it.

jaimecbernardo commented:
> I'm really looking forward to this. I am wondering if there should be a "Unlock file" option (likely using [`UnlockFileEx()`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-unlockfileex)). And, perhaps even a "Kill All" option?

Might be future improvements. Although those two seem dangerous 😅

jaimecbernardo commented:
@niels9001 , merging it so we can think about tests :) I think the icons can be included in a different PR and it's OK to ship with the default icons if needed.

jaimecbernardo commented:
Thanks for the reviews and your hard work, everyone!

niels9001 commented:
> @niels9001 , merging it so we can think about tests :) I think the icons can be included in a different PR and it's OK to ship with the default icons if needed.

👍.. wrapping those up, will PR those asap!

joedf commented:
> > I'm really looking forward to this. I am wondering if there should be a "Unlock file" option (likely using [`UnlockFileEx()`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-unlockfileex)). And, perhaps even a "Kill All" option?
> 
> Might be future improvements. Although those two seem dangerous 😅

I agree, but these could be put behind a confirm message box and have a little ⚠ icon.

Aaron-Junker commented:
@ivanstosic-janea Someone on Reddit asked me what the wallpaper in the OOBE image is. Do you know this?

ivanstosic-janea commented:
> @ivanstosic-janea Someone on Reddit asked me what the wallpaper in the OOBE image is. Do you know this?

Replied on Reddit :)

Aaron-Junker commented:
> > @ivanstosic-janea Someone on Reddit asked me what the wallpaper in the OOBE image is. Do you know this?
> 
> Replied on Reddit :)

You did?

I meant this comment: https://www.reddit.com/r/windows/comments/ykacrz/powertoys_release_064_with_two_new_utilities_file/iusn7wd?utm_medium=android_app&utm_source=share&context=3

ivanstosic-janea commented:
I did. I tried opening the thread in Incognito and my comment is not shown, I don't know why.
